crypto/odocrypt.c

```diff
@@ -0,0 +1,300 @@
+    {
+        uint64_t a = state[2*i];
+        uint64_t b = state[2*i+1];
+        // For each bit set in the mask, swap the corresponding bits in `a` and `b`
+        uint64_t swp = mask[i] & (a ^ b);
+        a ^= swp;
+        b ^= swp;
+        
+        state[2*i] = a;
+        state[2*i+1] = b;
+    }
+}
+
+void OdoCrypt_ApplyWordShuffle(uint64_t* state, int m)
+{
+    uint64_t next[STATE_SIZE];
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        next[m*i % STATE_SIZE] = state[i];
+    }
+    memcpy(state, &next[0], STATE_SIZE * sizeof(uint64_t));
+}
+
+void OdoCrypt_ApplyPboxRotations(uint64_t* state, const int* rotation)
+{
+    for (int i = 0; i < STATE_SIZE/2; i++)
+    {
+        // Only rotate the even words.  Rotating the odd words wouldn't actually
+        // be useful - a transformation that rotates all the words can be
+        // transformed into one that only rotates the even words, then rotates
+        // the odd words once after the final iteration.
+        state[2*i] = Rot(state[2*i], rotation[i]);
+    }
+}
+
+void OdoCrypt_ApplyPbox(uint64_t* state, const Pbox* perm)
+{
+    for (int i = 0; i < PBOX_SUBROUNDS-1; i++)
+    {
+        // Conditionally move bits between adjacent pairs of words
+        OdoCrypt_ApplyMaskedSwaps(state, perm->mask[i]);
+        // Move the words around
+        OdoCrypt_ApplyWordShuffle(state, PBOX_M);
+        // Rotate the bits within words
+        OdoCrypt_ApplyPboxRotations(state, perm->rotation[i]);
+    }
+    OdoCrypt_ApplyMaskedSwaps(state, perm->mask[PBOX_SUBROUNDS-1]);
+}
+
+void OdoCrypt_Pack(const uint64_t* state, char* bytes)
+{
+    memset(bytes, 0, ODOCRYPT_DIGEST_SIZE * sizeof(char));
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        for (int j = 0; j < 8; j++)
+        {
+            bytes[8*i + j] = (state[i] >> (8*j)) & 0xff;
+        }
+    }
+}
+
+void OdoCrypt_ApplySboxes(uint64_t* state, const uint8_t** sbox1, const uint16_t** sbox2)
+{
+    const static uint64_t MASK1 = (1 << SMALL_SBOX_WIDTH) - 1;
+    const static uint64_t MASK2 = (1 << LARGE_SBOX_WIDTH) - 1;
+    
+    int smallSboxIndex = 0;
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        uint64_t next = 0;
+        int pos = 0;
+        int largeSboxIndex = i;
+        for (int j = 0; j < SMALL_SBOX_COUNT / STATE_SIZE; j++)
+        {
+            next |= (uint64_t) sbox1[smallSboxIndex][(state[i] >> pos) & MASK1] << pos;
+            pos += SMALL_SBOX_WIDTH;
+            next |= (uint64_t) sbox2[largeSboxIndex][(state[i] >> pos) & MASK2] << pos;
+            pos += LARGE_SBOX_WIDTH;
+            smallSboxIndex++;
+        }
+        state[i] = next;
+    }
+}
+
+void OdoCrypt_ApplyRotations(uint64_t* state, const int* rotations)
+{
+    uint64_t next[STATE_SIZE];
+    
+    // rotate memcpy (rotate by one)
+    // @Mental: Please check this. Alternatively, I would achieve it with a loop.
+    // Original: std::rotate_copy(state, state+1, state+STATE_SIZE, next);
```

Comment:
If you're going to use memcpy you have to multiply the third argument by sizeof(uint64_t).  I would just do a loop:
```
for (int i = 1; i < STATE_SIZE; i++)
    next[i-1] = state[i];
next[STATE_SIZE-1] = state[0];
```

Reply:
Yeah, I actually know that (as you can see in the rest of the source code). Just forgot it here.
I prefer the loop as well :-)